lib/srv/discovery/eks_audit_log_fetcher.go

```diff
@@ -0,0 +1,192 @@
+	aws_sync "github.com/gravitational/teleport/lib/srv/discovery/fetchers/aws-sync"
+)
+
+// eksAuditLogFetcher is a fetcher for EKS audit logs for a single cluster,
+// fetching the logs from AWS Cloud Watch Logs. It uses the grpc stream
+// to initiate the stream and possibly receive a resume state used to
+// synchronize the start point with a previous run fetching the logs.
+type eksAuditLogFetcher struct {
+	fetcher *aws_sync.Fetcher
+	cluster *accessgraphv1alpha.AWSEKSClusterV1
+	stream  accessgraphv1alpha.AccessGraphService_KubeAuditLogStreamClient
+	log     *slog.Logger
+	cancel  context.CancelFunc
+}
+
+// Run starts continuously polling AWS Cloud Watch Logs for Kubernetes apiserver
+// audit logs for the configured cluster, feeding the logs retrieved to the
+// configured grpc stream. It runs until the given context is canceled.
+func (f *eksAuditLogFetcher) Run(ctx context.Context) error {
+	f.log = f.log.With("cluster", f.cluster.Arn)
+
+	cursor := initialCursor(f.cluster)
+	if err := f.sendTAGKubeAuditLogNewStream(ctx, cursor); err != nil {
+		return trace.Wrap(err)
+	}
+
+	cursor, err := f.receiveTAGKubeAuditLogResume(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	for ctx.Err() == nil {
+		var events []*structpb.Struct
+		events, cursor = f.fetchLogs(ctx, cursor)
+
+		if len(events) == 0 {
+			select {
+			case <-ctx.Done():
+				break
+			case <-time.After(logPollInterval):
+				continue
+			}
+		}
+
+		if err := f.sendTAGKubeAuditLogEvents(ctx, events, cursor); err != nil {
+			return trace.Wrap(err)
+		}
+
+		f.log.DebugContext(ctx, "Sent KubeAuditLogEvents", "count", len(events),
+			"cursor_time", cursor.LastEventTime.AsTime().Format(time.RFC3339))
+	}
+	return trace.Wrap(ctx.Err())
+}
+
+// fetchLogs fetches a batch of logs from AWS Cloud Watch Logs after the given
+// cursor position and unmarshals them into the protobuf Struct well-known
+// type. It returns those log entries and a cursor that can be used in the next
+// call to fetchLogs to get the logs after the ones returned. If an error
+// occurs, it is logged and no logs and the cursor passed in is returned. This
+// allows the called to continue to try to fetch logs until the fetcher is
+// terminated.
+func (f *eksAuditLogFetcher) fetchLogs(ctx context.Context, cursor *accessgraphv1alpha.KubeAuditLogCursor) ([]*structpb.Struct, *accessgraphv1alpha.KubeAuditLogCursor) {
+	awsEvents, err := f.fetcher.FetchEKSAuditLogs(ctx, f.cluster, cursor)
+	if err != nil {
+		if !errors.Is(err, context.Canceled) {
+			f.log.ErrorContext(ctx, "Failed to fetch EKS audit logs", "error", err)
+		}
+		return nil, cursor
+	}
+
+	if len(awsEvents) == 0 {
+		return nil, cursor
+	}
+
+	events := []*structpb.Struct{}
+	var awsEvent cwltypes.FilteredLogEvent
+	for _, awsEvent = range awsEvents {
+		// TODO(camscale): Track event sizes and don't go over protobuf message
+		// limit. newAccessGraphClient() sets the limit to 50MB
```

Comment:
Maybe add a note that FetchEKSAuditLogs limits to 500 iiuc, so unlikely to go over the 50MB